commit ea0fd4262e3b9a99865525b354d3a8da7e76afab (refs/changes/23/4623/2)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2018-08-07T18:07:33+00:00 (1 year, 2 months ago)
    
    TRITON-637 fwadm(1M) needs to add "keep state" to outbound allow rules
    Reviewed by: Jason King <jason.king@joyent.com>
    Approved by: Michael Zeller <mike.zeller@joyent.com>

diff --git a/src/fw/lib/fw.js b/src/fw/lib/fw.js
index 669fc836..1ee415f4 100644
--- a/src/fw/lib/fw.js
+++ b/src/fw/lib/fw.js
@@ -1259,7 +1259,8 @@ function prepareIPFdata(opts, log, callback) {
 
         conf[vm].sort(compareRules).forEach(function (sortObj) {
             var ktxt = KEEP_FRAGS;
-            if (sortObj.direction === 'to' && iks[sortObj.protocol]) {
+            if ((sortObj.direction === 'from' && sortObj.action === 'allow')
+                || (sortObj.direction === 'to' && iks[sortObj.protocol])) {
                 ktxt += KEEP_STATE;
             }
 
diff --git a/src/fw/test/lib/helpers.js b/src/fw/test/lib/helpers.js
index 22c7474d..45d4bf3d 100644
--- a/src/fw/test/lib/helpers.js
+++ b/src/fw/test/lib/helpers.js
@@ -696,24 +696,24 @@ function allowInICMP6(src, type, code) {
 }
 
 
-function allowPortOutTCP(dst, port) {
-    return createPortRule('pass', 'out', 'tcp', 'any to ' + dst, port);
+function allowPortOutTCP(dst, port, rest) {
+    return createPortRule('pass', 'out', 'tcp', 'any to ' + dst, port, rest);
 }
 
-function blockPortOutTCP(dst, port) {
-    return createPortRule('block', 'out', 'tcp', 'any to ' + dst, port);
+function blockPortOutTCP(dst, port, rest) {
+    return createPortRule('block', 'out', 'tcp', 'any to ' + dst, port, rest);
 }
 
-function allowPortOutUDP(dst, port) {
-    return createPortRule('pass', 'out', 'udp', 'any to ' + dst, port);
+function allowPortOutUDP(dst, port, rest) {
+    return createPortRule('pass', 'out', 'udp', 'any to ' + dst, port, rest);
 }
 
 function allowRangeOutUDP(dst, p1, p2) {
     return createRangeRule('pass', 'out', 'udp', 'any to ' + dst, p1, p2);
 }
 
-function blockPortOutUDP(dst, port) {
-    return createPortRule('block', 'out', 'udp', 'any to ' + dst, port);
+function blockPortOutUDP(dst, port, rest) {
+    return createPortRule('block', 'out', 'udp', 'any to ' + dst, port, rest);
 }
 
 function blockRangeOutUDP(dst, p1, p2) {
diff --git a/src/fw/test/unit/priority.test.js b/src/fw/test/unit/priority.test.js
index d31fc0a8..da2d8051 100644
--- a/src/fw/test/unit/priority.test.js
+++ b/src/fw/test/unit/priority.test.js
@@ -219,9 +219,9 @@ exports['Overriding outbound rules'] = function (t) {
             }, 'rules returned');
 
             v4rules[vm.uuid].out.udp = [
-                helpers.allowPortOutUDP('10.99.99.254', 22),
+                helpers.allowPortOutUDP('10.99.99.254', 22, 'keep state'),
                 helpers.blockRangeOutUDP('10.99.99.254', 15, 30),
-                helpers.allowPortOutUDP('10.99.99.0/24'),
+                helpers.allowPortOutUDP('10.99.99.0/24', '', 'keep state'),
                 helpers.blockPortOutUDP('any')
             ];
 
